accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,153 @@
+```
+
+This example contains three different queries:
+- Universal pillars looking for target `*`
+- Group pillars. This is done by looking up salt client `server_id` using its `minion_id` joining server group membership table to obtain `server_group_id` set where salt client is a member. Then looking up `target` in this result set.
+- Minion pillars looking for `minion_id`
+
+Salt automatically merge results from these queries, latest with the highest priority.
+
+### Upstream changes needed
+
+salt.pillar.postgres module, resp. salt.pillar.sql_base is not yet designed to work with JSON results directly from database and expects always to get results in *key:value* format. To enable JSON output, small patch need to be negotiated with upstream (patch against pre-blackened version):
+
+```diff
+diff --git a/salt/pillar/sql_base.py b/salt/pillar/sql_base.py
+index eeb2269f77..dae0b36204 100644
+--- a/salt/pillar/sql_base.py
++++ b/salt/pillar/sql_base.py
+@@ -352,8 +352,13 @@ class SqlBaseExtPillar(six.with_metaclass(abc.ABCMeta, object)):
+                         # dict, descend.
+                         crd = crd[ret[i]]
+
++            # We have just one field without any key, assume returned row is already a dict
++            # aka JSON storage
++            if self.num_fields == 1:
++                crd.update(ret[0])
++
+             # If this test is true, the penultimate field is the key
+-            if self.depth == self.num_fields - 1:
++            elif self.depth == self.num_fields - 1:
+                 nk = self.num_fields-2  # Aka, self.depth-1
+                 # Should we and will we have a list at the end?
+                 if ((self.as_list and (ret[nk] in crd)) or
+```
+
+## Maintaining consistency
+
+Responsibility to provide up to date data remains, as is the case today, on the component changing the data. That means, when information about minion is changed, then relevant pillar generator must be called to refresh the pillar in database.
+However with the pillar in database and loosely (no foreign key relation) connected to minions and groups, database trigger which can remove pillar entries when minion or group is removed:
+
+```sql
+CREATE OR REPLACE FUNCTION minion_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.minion_id;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE FUNCTION group_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.id::VARCHAR;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE TRIGGER minion_removed BEFORE DELETE ON suseMinionInfo FOR EACH ROW EXECUTE PROCEDURE minion_removed();
+
+CREATE TRIGGER group_removed BEFORE DELETE ON rhnServerGroup FOR EACH ROW EXECUTE PROCEDURE group_removed();
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we **not** do this?
+
+* will it impact performance?
+  
+  TBD, I suspect better performance then with thousands of files in one directory
```

Comment:
Is it possible to test this? In particular, I wonder what happens if a lot of pillars are requested in a short amount of time (eg. highstate application on many minions concurrently). Will each pillar request need an opening of a new DB connection, or are those pooled somehow? Is there a caching mechanism of any sort?

Reply:
I don't think there is caching of pillars for highstate and other states applications (see [in-memory vs on-demand pillars](https://docs.saltproject.io/en/latest/topics/pillar/#in-memory-pillar-data-vs-on-demand-pillar-data)). Also I didn't see any batching in the code of external pillar and seems it is all designed for per-minion queries.
As I proposed in the external pillar settings, it is 3 DB queries for a minion so in case of e.g. 10000 minions, we will get 30000 queries concurrently + other queries resulting from state applications.

Compare to current state of `suma-minion.py` external pillar. That means 10000 executions of external pillar, accessing from 2 to many files (formula, image, virtualization, custom info pillars).

To be fair, in the initial implementation I plan to migrate pillars generated by `MinionPillarManager` and `SaltStateGeneratorService::generateImagePillar`. This does not cover formulas yet as there it is a little more work. So `suma-minion.py` will need to stay for time being until formulas are migrated as well.

As for testing, I should be able to gather some performance data using evil minions. I will look into it.

Reply:
I did a quick test with evil minions:

2 regular minions, 100 evil minions

Command:
time bash -c 'for i in `seq 1 100`;do salt "*" saltutil.pillar_refresh > /dev/null; done'

* for 102 minions and suma-minion.py only:

```
real    4m37,656s
user    1m53,964s
sys     0m10,080s
```


* for 102 minions and psql pillar only:

```
real    4m39,619s
user    1m54,091s
sys     0m10,101s
```


* for 102 minions and both suma-minion.py and psql pillar:

```
real    4m40,897s
user    1m52,634s
sys     0m10,143s
```

I was monitoring CPU load on the server and majority was taken by salt, (~30%). For postgresql this wasn't difficult at all steadily remaining under 1%.